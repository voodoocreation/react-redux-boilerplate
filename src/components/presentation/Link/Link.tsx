import NextLink from "next/link";
import { withRouter } from "next/router";
import * as React from "react";

import Routes from "../../../../next.routes";

export interface IProps extends React.HTMLAttributes<HTMLAnchorElement> {
  href?: string;
  isExternal?: boolean;
  params?: {};
  prefetch?: boolean;
  route?: string;
}

const Link: React.FC<IProps & { router: any }> = ({
  children,
  href,
  isExternal,
  params,
  prefetch,
  route,
  router,
  ...props
}) => {
  const externalAttrs: { target?: string; rel?: string } = {};

  if (isExternal) {
    externalAttrs.target = "_blank";
    externalAttrs.rel = "noopener noreferrer";
  }

  if (!href && !route) {
    return <span {...props}>{children}</span>;
  }

  if (
    !router ||
    !router.components ||
    Object.keys(router.components).length < 1
  ) {
    return (
      <a href={route || href} {...externalAttrs} {...props}>
        {children}
      </a>
    );
  }

  if (route) {
    return (
      <Routes.Link params={params} prefetch={prefetch} route={route}>
        <a {...externalAttrs} {...props}>
          {children}
        </a>
      </Routes.Link>
    );
  }

  return (
    <NextLink href={href!} prefetch={prefetch}>
      <a {...externalAttrs} {...props}>
        {children}
      </a>
    </NextLink>
  );
};

Link.displayName = "Link";
Link.defaultProps = {
  isExternal: false,
  prefetch: false
};

export default withRouter(Link);
