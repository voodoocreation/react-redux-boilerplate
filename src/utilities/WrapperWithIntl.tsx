import * as React from "react";
import { IntlConfig } from "react-intl";
import { WrapperWithIntl as BaseWrapper } from "react-test-wrapper";

import messages from "../locales/en-NZ";

export default class WrapperWithIntl<
  C extends React.ComponentType<any>,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> extends BaseWrapper<C, P> {
  protected intlProviderProps: Partial<IntlConfig> = {
    messages
  };
}
