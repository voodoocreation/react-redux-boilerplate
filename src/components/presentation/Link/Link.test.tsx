import WrapperWithIntl from "../../../utilities/WrapperWithIntl";
import Link from "./Link";

const component = new WrapperWithIntl(Link)
  .withDefaultProps({
    router: {
      components: {
        test: ""
      }
    }
  } as any)
  .withDefaultChildren("Link text");

describe("[presentation] <Link />", () => {
  describe("when the router is available and route is defined", () => {
    const wrapper = component.withProps({ route: "/" }).mount();

    it("renders with LinkRoutes", () => {
      expect(wrapper.find("LinkRoutes")).toHaveLength(1);
    });

    it("renders with Next.js Link", () => {
      expect(wrapper.find("Link")).toHaveLength(2);
    });

    it("renders an <a>", () => {
      expect(wrapper.find("a")).toHaveLength(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when the router is available and href is defined", () => {
    const wrapper = component.withProps({ href: "/" }).mount();

    it("doesn't render with LinkRoutes", () => {
      expect(wrapper.find("LinkRoutes")).toHaveLength(0);
    });

    it("renders with Next.js Link", () => {
      expect(wrapper.find("Link")).toHaveLength(2);
    });

    it("renders an <a>", () => {
      expect(wrapper.find("a")).toHaveLength(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when the router is unavailable and href is defined", () => {
    const wrapper = component
      .withProps({
        href: "/",
        router: null
      } as any)
      .mount();

    it("doesn't render with LinkRoutes", () => {
      expect(wrapper.find("LinkRoutes")).toHaveLength(0);
    });

    it("doesn't render with Next.js Link", () => {
      expect(wrapper.find("Link")).toHaveLength(1);
    });

    it("renders an <a>", () => {
      expect(wrapper.find("a")).toHaveLength(1);
    });

    it("<a> href is the href value", () => {
      expect(wrapper.find("a").prop("href")).toBe("/");
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when isExternal is true", () => {
    const wrapper = component
      .withProps({
        isExternal: true,
        route: "/",
        router: null
      } as any)
      .mount();

    it("external attributes are defined on the <a>", () => {
      expect(wrapper.find("a").props()).toMatchObject({
        rel: "noopener noreferrer",
        target: "_blank"
      });
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when both href and route are undefined", () => {
    const wrapper = component
      .withProps({
        href: undefined,
        route: undefined
      })
      .mount();

    it("doesn't render an <a>", () => {
      expect(wrapper.find("a")).toHaveLength(0);
    });

    it("renders a <span>", () => {
      expect(wrapper.find("span")).toHaveLength(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
