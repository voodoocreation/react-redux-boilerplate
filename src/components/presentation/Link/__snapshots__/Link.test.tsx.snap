// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Link /> when both href and route are undefined matches snapshot 1`] = `
<span>
  Link text
</span>
`;

exports[`[presentation] <Link /> when isExternal is true matches snapshot 1`] = `
<a
  href="/"
  rel="noopener noreferrer"
  target="_blank"
>
  Link text
</a>
`;

exports[`[presentation] <Link /> when the router is available and href is defined matches snapshot 1`] = `
<a
  href="/"
>
  Link text
</a>
`;

exports[`[presentation] <Link /> when the router is available and route is defined matches snapshot 1`] = `
<a
  href="/"
>
  Link text
</a>
`;

exports[`[presentation] <Link /> when the router is unavailable and href is defined matches snapshot 1`] = `
<a
  href="/"
>
  Link text
</a>
`;
